### PR TITLE
Always report the expanded location for a declaration

### PR DESCRIPTION
**Pull request by jru**
Hi,

I think I found an elegant solution to the problem of reporting the proper location for declarations that were generated by macros.

Let me know what you think and if I overlooked something. Thanks!

Related: #282

### CODE REVIEW

**Comment by kimgr** [comment]
I spent some time on macro reporting a while back, and there are a few cases that seem tricky to handle correctly:

- Macro arguments -- they should usually be attributed to the macro instantiation loc, you should be responsible for stuff you pass in
- Scratch space/macro arg concatenation -- if the symbol is ultimately formed inside the macro definition, but the result of concatenating macro arguments, it should be attributed to the instantiation loc, see above
 
See `GetCanonicalUseLocation` for more details:
https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu.cc#L1366

It could be that your solution handles this, and we can simplify `GetCanonicalUseLocation` too, but I spent some considerable effort coming up with this and fixing things like nested macro expansions, so I'm kind-of hoping it doesn't :-p

**Comment by jru** [comment]
Hi,
I hope not to disappoint you, then :)

Well, if I understood you right, from your points I only solved the second (the first already works). Let me explain:

Case 1: If a declaration lives oustide of a macro definition body, any use of it will point to a valid location, no matter if the use was materialized from a macro or spelled directly:
```cpp
// file a.h
struct A {};
// ----------

// file main.cc
#include "a.h"
#define MACRO(p) A p

A p1;       // Full use of A, which is in file a.h
MACRO(p2);  // Full use of A, which is still in file a.h
// ----------
```
In this case, the declaration *A* lives outside of any macro and it should always be reported as coming from *a.h* independently of where the use is. In this case, my fix will do nothing (since the spelling and instantiaton location of the decl *A* are the same, right?)

----
    
Case 2: If the declaration lives inside of the macro definition and a use refers to the expanded tokens (from the macro body), the declaration should point to whatever file contains the invocation (given that the user wouldn't know anything about the macro):

```cpp
// file macro.h
#define GEN(x) void name_ ## x ()
#define CALL(obj, method) obj.method();
// ----------

// file a.h
#include "macro.h"
struct A { GEN(blah); };
// ----------

// file main.cc
#include "a.h"
A p1;                // Full use of A, which is in file a.h
p1.name_blah();      // Full use of A::name_blah, which is in file a.h (expanded location from GEN)
CALL(p2, name_blah); // Full use of A::name_blah, but this time inside a macro. Need to include its definition from macro.h
// ----------
```
In this case, the first call to name_blah() needs the expanded location since we use the symbol directly. The second call, however, is wrapped in a macro so we need to include its definition, in *macro.h*. My fix makes sure that these two calls get the proper location.

----

I think there is another case, probably much harder to solve (and much more interesting :)) which is: what happens to dependent symbols present in the macro definition but not included directly by it? For example, the example you posted in the issue #283.  Another example would be the case 1 if the macro had lived in its own file without including anything. In that case, should it be responsible of including the decl of A? Not sure. 

I have the feeling all of this is closely related to solving re-export issue you mentioned in other posts.


Please take my assumptions with a grain of salt. I am pretty new to the project and I am still wrapping my head around the code. Please, do let me know if I am completely off :)

Cheers

**Comment by kimgr** [comment]
Thanks for the walkthrough.

I played with this a little just now. One thing that surprised me was that the test only covers half the conditional in `GetDeclFileEntryForUse` -- I added logging like this:

    clang::SourceLocation decl_loc = GetLocation(decl);
    if (use.isMacroID()) {
     VERRS(3) << "Use of '" << PrintableDecl(decl) << "' at "
              << PrintableLoc(use) << " is a macro expansion, using decl loc: "
              << PrintableLoc(decl_loc) << "\n";
    } else {
       decl_loc = GetInstantiationLoc(decl_loc);
       VERRS(3) << "Use of '" << PrintableDecl(decl) << "' at "
                << PrintableLoc(use) << " is an explicit spelling, using decl loc: "
                << PrintableLoc(decl_loc) << "\n";
    }

I could only get the else-leg to run, which means we can reduce this function to:

    clang::SourceLocation decl_loc = GetInstantiationLoc(GetLocation(decl));
    return GetLocFileEntry(decl_loc);

which is probably not right.

And I think the original logic is slightly off -- it checks if `use` is a macro expansion, but I don't think that says anything about whether we should point to the decl's splocation-in-macro or location-as-expanded.

**Comment by jru** [comment]
Wow, nice catch! You are completely right..

I assumed that the source location we get for the use will be the raw location, but now I see that it is converted in GetCanonicalUseLocation() to either, the expansion, or the spelling location and they are always file IDs (right?)

This will definitely need to be fixed. I will need some time to understand what we are really doing here. 
Thanks for pointing this out!

**Comment by jru** [comment]
It turns out the solution is exactly what you mentionied in your comment: The decl location should always be the expanded location.  So simple it is hard to believe :)

The explanation for that is that the use shouldn't really care how the decl was generated, however it is, it should always point to the file that ultimately declares it. In turn, that include file is responsible for including whatever macros it needs itself but that should in no case leak where the use is. Makes sense?

**Comment by kimgr** [comment]
I haven't been able to get my head fully around this, but I think I agree -- a declaration should always be attributed to the expansion location.

Could you add a test case for nested expansions as well? I.e. a macro that expands to a class declaration containing macro that expands to a method declaration.

I've had trouble with nested macros before, but I think maybe the same reasoning holds there -- it's only the ultimate expansion location we're interested in, right?

**Comment by jru** [comment]
Hi again,

I wrote the extra test you suggested and while running it I found out that there are still several places in the code where a declaration is incorrectly attributed to the spelling location. So I guess there is more work to do here..

Concretely, whenever *GetFileEntry* or *GetFilePath* is called with a decl or a macro location, the spelling location will be always returned. This is the case in most of the rules checked during *ProcessFullUse* but I have seen it also in other places.

It is hard for me to tell whether all of those locations should be always turn into the "instantiation" location, so I am kind of in a blocked state right now.

What do you think? Should I try to get this through with the minimal changes now and thing about refactoring the GetLocation/GetFileEntry/GetFilePath in the future, or do you think it makes sense to get this straight right now?

**Comment by jru** [comment]
I did some more digging using clang-query to find all the exact places where the location of a decl is used and I got 67 matches. 

Excluding those only used to print the location, the rest deserves a closer look. I am sure the behavior will not be the intended one for decls coming from macro expansions defined in different files.


The matcher I used:
```
let locationDecls functionDecl(anyOf(hasName("GetLocation"), hasName("GetFileEntry"), hasName("GetFilePath")))
let declPointer pointerType(pointee(hasDeclaration(cxxRecordDecl(isSameOrDerivedFrom(hasName("clang::Decl"))))))
m callExpr(callee(locationDecls), hasArgument(0, hasType(declPointer)))
```

And the result of running it through all iwyu files:
```

clang-query -p=.\ -f=matcher.txt iwyu.cc iwyu_ast_util.cc iwyu_cache.cc iwyu_driver.cc iwyu_getopt.cc iwyu_globals.cc iwyu_include_picker.cc iwyu_lexer_utils.cc iwyu_location_util.cc iwyu_output.cc iwyu_path_util.cc iwyu_preprocessor.cc iwyu_verrs.cc  

Match #1:

iwyu.cc(1391,11):  note: "root" binds here
      if (GetFileEntry(redecl) == macro_def_file && IsForwardDecl(redecl)) {
          ^~~~~~~~~~~~~~~~~~~~

Match #2:

iwyu.cc(1490,38):  note: "root" binds here
              GetFileEntry(use_loc), GetFileEntry(dfn))) {
                                     ^~~~~~~~~~~~~~~~~

Match #3:

iwyu.cc(1547,56):  note: "root" binds here
    if (CodeAuthorWantsJustAForwardDeclare(deref_type, GetLocation(decl))) {
                                                       ^~~~~~~~~~~~~~~~~

Match #4:

iwyu.cc(1618,57):  note: "root" binds here
    if (CodeAuthorWantsJustAForwardDeclare(return_type, GetLocation(decl))) {
                                                        ^~~~~~~~~~~~~~~~~

Match #5:

iwyu.cc(1838,56):  note: "root" binds here
    if (CodeAuthorWantsJustAForwardDeclare(deref_type, GetLocation(decl)) ||
                                                       ^~~~~~~~~~~~~~~~~

Match #6:

iwyu.cc(1883,61):  note: "root" binds here
           !CodeAuthorWantsJustAForwardDeclare(return_type, GetLocation(decl)));
                                                            ^~~~~~~~~~~~~~~~~

Match #7:

iwyu.cc(1892,32):  note: "root" binds here
      ReportTypeUseWithComment(GetLocation(decl), return_type,
                               ^~~~~~~~~~~~~~~~~

Match #8:

iwyu.cc(2368,46):  note: "root" binds here
    const FileEntry* first_decl_file_entry = GetFileEntry(first_decl);
                                             ^~~~~~~~~~~~~~~~~~~~~~~~

Match #9:

iwyu.cc(2371,11):  note: "root" binds here
      if (GetFileEntry(*it) != first_decl_file_entry)
          ^~~~~~~~~~~~~~~~~

Match #10:

iwyu.cc(2720,11):  note: "root" binds here
      if (GetFileEntry(decl) == GetFileEntry(using_decl) || // prefer same file
          ^~~~~~~~~~~~~~~~~~

Match #11:

iwyu.cc(2720,33):  note: "root" binds here
      if (GetFileEntry(decl) == GetFileEntry(using_decl) || // prefer same file
                                ^~~~~~~~~~~~~~~~~~~~~~~~

Match #12:

iwyu.cc(1391,11):  note: "root" binds here
      if (GetFileEntry(redecl) == macro_def_file && IsForwardDecl(redecl)) {
          ^~~~~~~~~~~~~~~~~~~~

Match #13:

iwyu.cc(1490,38):  note: "root" binds here
              GetFileEntry(use_loc), GetFileEntry(dfn))) {
                                     ^~~~~~~~~~~~~~~~~

Match #14:

iwyu.cc(1547,56):  note: "root" binds here
    if (CodeAuthorWantsJustAForwardDeclare(deref_type, GetLocation(decl))) {
                                                       ^~~~~~~~~~~~~~~~~

Match #15:

iwyu.cc(1618,57):  note: "root" binds here
    if (CodeAuthorWantsJustAForwardDeclare(return_type, GetLocation(decl))) {
                                                        ^~~~~~~~~~~~~~~~~

Match #16:

iwyu.cc(1838,56):  note: "root" binds here
    if (CodeAuthorWantsJustAForwardDeclare(deref_type, GetLocation(decl)) ||
                                                       ^~~~~~~~~~~~~~~~~

Match #17:

iwyu.cc(1883,61):  note: "root" binds here
           !CodeAuthorWantsJustAForwardDeclare(return_type, GetLocation(decl)));
                                                            ^~~~~~~~~~~~~~~~~

Match #18:

iwyu.cc(1892,32):  note: "root" binds here
      ReportTypeUseWithComment(GetLocation(decl), return_type,
                               ^~~~~~~~~~~~~~~~~

Match #19:

iwyu.cc(2368,46):  note: "root" binds here
    const FileEntry* first_decl_file_entry = GetFileEntry(first_decl);
                                             ^~~~~~~~~~~~~~~~~~~~~~~~

Match #20:

iwyu.cc(2371,11):  note: "root" binds here
      if (GetFileEntry(*it) != first_decl_file_entry)
          ^~~~~~~~~~~~~~~~~

Match #21:

iwyu.cc(2720,11):  note: "root" binds here
      if (GetFileEntry(decl) == GetFileEntry(using_decl) || // prefer same file
          ^~~~~~~~~~~~~~~~~~

Match #22:

iwyu.cc(2720,33):  note: "root" binds here
      if (GetFileEntry(decl) == GetFileEntry(using_decl) || // prefer same file
                                ^~~~~~~~~~~~~~~~~~~~~~~~

Match #23:

iwyu.cc(1391,11):  note: "root" binds here
      if (GetFileEntry(redecl) == macro_def_file && IsForwardDecl(redecl)) {
          ^~~~~~~~~~~~~~~~~~~~

Match #24:

iwyu.cc(1490,38):  note: "root" binds here
              GetFileEntry(use_loc), GetFileEntry(dfn))) {
                                     ^~~~~~~~~~~~~~~~~

Match #25:

iwyu.cc(1547,56):  note: "root" binds here
    if (CodeAuthorWantsJustAForwardDeclare(deref_type, GetLocation(decl))) {
                                                       ^~~~~~~~~~~~~~~~~

Match #26:

iwyu.cc(1618,57):  note: "root" binds here
    if (CodeAuthorWantsJustAForwardDeclare(return_type, GetLocation(decl))) {
                                                        ^~~~~~~~~~~~~~~~~

Match #27:

iwyu.cc(1838,56):  note: "root" binds here
    if (CodeAuthorWantsJustAForwardDeclare(deref_type, GetLocation(decl)) ||
                                                       ^~~~~~~~~~~~~~~~~

Match #28:

iwyu.cc(1883,61):  note: "root" binds here
           !CodeAuthorWantsJustAForwardDeclare(return_type, GetLocation(decl)));
                                                            ^~~~~~~~~~~~~~~~~

Match #29:

iwyu.cc(1892,32):  note: "root" binds here
      ReportTypeUseWithComment(GetLocation(decl), return_type,
                               ^~~~~~~~~~~~~~~~~

Match #30:

iwyu.cc(2368,46):  note: "root" binds here
    const FileEntry* first_decl_file_entry = GetFileEntry(first_decl);
                                             ^~~~~~~~~~~~~~~~~~~~~~~~

Match #31:

iwyu.cc(2371,11):  note: "root" binds here
      if (GetFileEntry(*it) != first_decl_file_entry)
          ^~~~~~~~~~~~~~~~~

Match #32:

iwyu.cc(2720,11):  note: "root" binds here
      if (GetFileEntry(decl) == GetFileEntry(using_decl) || // prefer same file
          ^~~~~~~~~~~~~~~~~~

Match #33:

iwyu.cc(2720,33):  note: "root" binds here
      if (GetFileEntry(decl) == GetFileEntry(using_decl) || // prefer same file
                                ^~~~~~~~~~~~~~~~~~~~~~~~

Match #34:

iwyu.cc(3139,15):  note: "root" binds here
              GetFileEntry(decl)))
              ^~~~~~~~~~~~~~~~~~

Match #35:

iwyu.cc(3390,17):  note: "root" binds here
                GetFileEntry(tpl_decl), GetFileEntry(resugared_decl)))
                ^~~~~~~~~~~~~~~~~~~~~~

Match #36:

iwyu.cc(3390,41):  note: "root" binds here
                GetFileEntry(tpl_decl), GetFileEntry(resugared_decl)))
                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~

Match #37:

iwyu.cc(3564,29):  note: "root" binds here
      if (CanIgnoreLocation(GetLocation(fd)))
                            ^~~~~~~~~~~~~~~

Match #38:

iwyu_ast_util.cc(196,14):  note: "root"
      binds here
      *loc = GetLocation(as_decl_);   // in iwyu_location_util.h
             ^~~~~~~~~~~~~~~~~~~~~

Match #39:

iwyu_ast_util.cc(901,22):  note: "root"
      binds here
      !IsBuiltinFile(GetFileEntry(decl)))
                     ^~~~~~~~~~~~~~~~~~

Match #40:

iwyu_ast_util.cc(1015,32):  note: "root"
      binds here
  FullSourceLoc first_decl_loc(GetLocation(first_decl), *GlobalSourceManager());
                               ^~~~~~~~~~~~~~~~~~~~~~~

Match #41:

iwyu_ast_util.cc(1021,36):  note: "root"
      binds here
    const FullSourceLoc redecl_loc(GetLocation(redecl), *GlobalSourceManager());
                                   ^~~~~~~~~~~~~~~~~~~

Match #42:

iwyu_output.cc(273,18):  note: "root" binds
      here
    decl_file_ = GetFileEntry(decl);
                 ^~~~~~~~~~~~~~~~~~

Match #43:

iwyu_output.cc(274,22):  note: "root" binds
      here
    decl_filepath_ = GetFilePath(decl);
                     ^~~~~~~~~~~~~~~~~

Match #44:

iwyu_output.cc(557,29):  note: "root" binds
      here
    decl_loc = PrintableLoc(GetLocation(use.decl()));
                            ^~~~~~~~~~~~~~~~~~~~~~~

Match #45:

iwyu_output.cc(686,7):  note: "root" binds
      here
  if (GetFileEntry(decl) != GetFileEntry(use.use_loc()))
      ^~~~~~~~~~~~~~~~~~

Match #46:

iwyu_output.cc(696,11):  note: "root" binds
      here
          GetLocation(decl) == use.use_loc() ||
          ^~~~~~~~~~~~~~~~~

Match #47:

iwyu_output.cc(1009,41):  note: "root" binds
      here
    if (GetFileEntry(use->use_loc()) != GetFileEntry(use->decl())) {
                                        ^~~~~~~~~~~~~~~~~~~~~~~~~

Match #48:

iwyu_output.cc(1024,40):  note: "root" binds
      here
    const SourceLocation defined_loc = GetLocation(redecl);
                                       ^~~~~~~~~~~~~~~~~~~

Match #49:

iwyu_output.cc(1053,39):  note: "root" binds
      here
  if (GetFileEntry(use->use_loc()) == GetFileEntry(use->decl()) &&
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~

Match #50:

iwyu_output.cc(1093,32):  note: "root" binds
      here
               << PrintableLoc(GetLocation(use->decl())) << "\n";
                               ^~~~~~~~~~~~~~~~~~~~~~~~

Match #51:

iwyu_output.cc(1109,29):  note: "root" binds
      here
    const string dfn_file = GetFilePath(fn_decl);
                            ^~~~~~~~~~~~~~~~~~~~

Match #52:

iwyu_output.cc(1132,49):  note: "root" binds
      here
        = picker.GetCandidateHeadersForFilepath(GetFilePath(method_dfn));
                                                ^~~~~~~~~~~~~~~~~~~~~~~

Match #53:

iwyu_output.cc(1134,49):  note: "root" binds
      here
        = picker.GetCandidateHeadersForFilepath(GetFilePath(parent_dfn));
                                                ^~~~~~~~~~~~~~~~~~~~~~~

Match #54:

iwyu_output.cc(1140,20):  note: "root" binds
      here
      same_file = (GetFileEntry(method_dfn) == GetFileEntry(parent_dfn));
                   ^~~~~~~~~~~~~~~~~~~~~~~~

Match #55:

iwyu_output.cc(1140,48):  note: "root" binds
      here
      same_file = (GetFileEntry(method_dfn) == GetFileEntry(parent_dfn));
                                               ^~~~~~~~~~~~~~~~~~~~~~~~

Match #56:

iwyu_output.cc(1161,11):  note: "root" binds
      here
          GetFileEntry(use->decl()), GetFileEntry(use->use_loc()))) {
          ^~~~~~~~~~~~~~~~~~~~~~~~~

Match #57:

iwyu_output.cc(1174,43):  note: "root" binds
      here
            GetFileEntry(use->use_loc()), GetFileEntry(use->decl()))) {
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~

Match #58:

iwyu_output.cc(1258,11):  note: "root" binds
      here
          GetFilePath(dfn), GetFilePath(use->use_loc()));
          ^~~~~~~~~~~~~~~~

Match #59:

iwyu_output.cc(1287,44):  note: "root" binds
      here
      if (ContainsKey(associated_includes, GetFileEntry(redecl))) {
                                           ^~~~~~~~~~~~~~~~~~~~

Match #60:

iwyu_output.cc(1300,30):  note: "root" binds
      here
             << PrintableLoc(GetLocation(providing_decl)) << "\n";
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~

Match #61:

iwyu_output.cc(1302,25):  note: "root" binds
      here
    const string file = GetFilePath(dfn);
                        ^~~~~~~~~~~~~~~~

Match #62:

iwyu_output.cc(1309,30):  note: "root" binds
      here
             << PrintableLoc(GetLocation(providing_decl)) << "\n";
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~

Match #63:

iwyu_output.cc(1312,9):  note: "root" binds
      here
    if (GetFileEntry(same_file_decl) != GetFileEntry(use->use_loc()))
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~

Match #64:

iwyu_output.cc(1313,33):  note: "root" binds
      here
      use->set_suggested_header(GetFilePath(same_file_decl));
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~

Match #65:

iwyu_output.cc(1334,30):  note: "root" binds
      here
             << PrintableLoc(GetLocation(dfn)) << "\n";
                             ^~~~~~~~~~~~~~~~

Match #66:

iwyu_output.cc(1338,30):  note: "root" binds
      here
             << PrintableLoc(GetLocation(same_file_decl)) << "\n";
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~

Match #67:

iwyu_output.cc(1684,20):  note: "root" binds
      here
      file_entry = GetFileEntry(dfn);
                   ^~~~~~~~~~~~~~~~~
67 matches.
```

**Comment by kimgr** [comment]
OK, that's a little overwhelming...

I think if you have a specific problem that you can solve with a relatively isolated patch, it's better to do that. Your current patch has the downside that it breaks the `badinc.cc` test case:

    AssertionError: 
    tests/cxx/badinc.cc:617: Unmatched regex:
    I1_TemplateClass is...*badinc-i1.h

I think that's because we currently just call `GetFileEntry(decl)`, which is the equivalent of `GetFileEntry(GetLocation(decl))`. The `GetLocation` overload for `Decl` handles a couple of special cases [1], and I think this test covers one of them.

So by using `decl->getLocation()` instead of `GetLocation(decl)`, you've missed out on that special-casing. Not sure if the macro-instantiation logic works if you switch to the latter?

Are you developing this on Windows? Unfortunately we've never got the test-suite 100% successful there, so it's hard to spot errors like these.

[1] https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_location_util.cc#L39

**Comment by jru** [comment]
Yes, I was testing on Windows only. I set up now a linux environment and will be always double checking there. Thanks for that catch, again.. :)

You are right, we can't expect to fix all cases with one patch. I corrected just enough to make the new test case pass. It is a bit more code but it should be clear what is going on.

**Comment by kimgr** [comment]
Thanks for the update! Tests all pass now on my Linux box.

Can you add license headers and format the tests to match general code style (2 spaces indentation, primarily). You can use clang-format to automatically format the new files to match prevailing style.

I'll walk through this in more detail when I have a little time. I wonder if we could add a test based on that other bug to see if it's covered as well...

**Comment by jru** [comment]
I corrected the formatting and added the LICENSE headers are you suggested.
What other bug do you mean?

**Comment by kimgr** [comment]
#282

**Comment by jru** [comment]
Good point! The fix covers it as well. 
It is actually the same case I am testing, since the template doesn't play a role here.

Proof :)
```
$ cat baseclas.h myclass.h mytemplate.h 282.cc
// baseclass.h
#define BASE_MACRODEF(NAME)  \
    static const char* GetName() { return #NAME; }

class CBaseClass {
public:
    BASE_MACRODEF(CBaseClass)
};
// myclass.h
#include "baseclass.h"

class CMyClass : public CBaseClass {
public:
    BASE_MACRODEF(CMyClass)
};

// mytemplate.h
template <class T>
class CMyTemplate {
public:
    CMyTemplate() : m_name(T::GetName()) {}

private:
    const char* m_name;
};

// test.cpp
#include "myclass.h"
#include "mytemplate.h"

int main(int argc, char* argv[]) {
    CMyTemplate<CMyClass> myTemplate;
    return 0;
}
$ ~/build-llvm/bin/include-what-you-use 282.cc

(282.cc has correct #includes/fwd-decls)
```

File: iwyu_output.cc [hunk 1]
```diff
@@ -1120,6 +1121,11 @@ void ProcessFullUse(OneUse* use,
     // See if we also recorded a use of the parent.
     const NamedDecl* parent_dfn
         = GetDefinitionAsWritten(method_dfn->getParent());
+
+    auto const decl_file_entry = GetFileEntry(use->decl_loc());
```
[the remark this answers is not in the document]
Comment (jru) [reply iwyu_output.cc]: Done

**Comment by kimgr** [comment]
I found a typo in the commit message ('...where the it was ultimately...') that I fixed up locally. Otherwise your change went in unmodified. Thank you for taking this on, this was a long-standing bug!

**Pull request closed without merging**